src/prefect/engine/cloud/flow_runner.py

```diff
@@ -130,10 +134,104 @@ def call_runner_target_handlers(self, old_state: State, new_state: State) -> Sta
             state.state = old_state  # type: ignore
             raise ENDRUN(state=state)
 
+        if isinstance(state, Cancelled):
+            finished_states = [s.__name__ for s in Finished.children()]
+            query = {
+                "query": {
+                    with_args("flow_run_by_pk", {"id": flow_run_id}): {
+                        with_args(
+                            "task_runs",
+                            {"where": {"state": {"_nin": finished_states}}},
+                        ): {"id": True, "state": True}
+                    }
+                }
+            }
+            res = self.client.graphql(query)
+            if res.data.flow_run_by_pk.task_runs:
+                task_state = Cancelled("Flow run cancelled").serialize()
+                mutation = {
+                    "mutation($input: set_task_run_states_input!)": {
+                        "set_task_run_states(input: $input)": {
+                            "states": {"id", "status", "message"}
+                        }
+                    }
+                }
+                variables = {
+                    "input": {
+                        "states": [
+                            {"state": task_state, "task_run_id": run.id}
+                            for run in res.data.flow_run_by_pk.task_runs
+                        ]
+                    }
+                }
+                self.client.graphql(mutation, variables=variables)  # type: ignore
```

Comment:
My initial reaction to this code block is that setting the individual task runs to `Cancelled` should be the responsibility of the backend not the client - do we know at this moment that no task runs are actively running and might report back a state update?

Reply:
That was my initial reaction as well, but it causes issues with users running past versions of `prefect`.

Currently the "cancellation lite" mechanism lets users set the state of a flow run to `Cancelled`. This prevents new tasks from entering `Running`, but allows existing tasks to continue. Changing the transition to `Cancelled` to also move all non-finished tasks to `Cancelled` on the backend side would break this for users running older versions of prefect, since tasks that are still running would be marked as `Cancelled`.

Reply:
New idea: Perhaps on transition to a `Cancelled` state, cloud sets all non-`Running` task runs to `Cancelled`. This does what we want for new versions of core (there will never be `Running` task runs at this point, so all remaining task runs will be cancelled), but doesn't affect older versions of core (running tasks will still be marked as running).

Reply:
> Changing the transition to Cancelled to also move all non-finished tasks to Cancelled on the backend side would break this for users running older versions of prefect, since tasks that are still running would be marked as Cancelled.

So what I envisioned was:
- user hits "cancel" 
- flow is put in "Cancelling" state (if it's running)
- executor attempts to cancel things and eventually sets Flow to `Cancelled`
- Zombie killer service finds task runs / flow runs that haven't heartbeated and does some state-logic to determine whether to set task runs / flow runs to Cancelled

I'd rather not aggressively set task run states to `Cancelled`, only because if they do actually finish successfully it's more informative for their state to reflect that.

Reply:
The zombie killer service enhancement I mentioned here is still compatible with Core setting task runs to `Cancelled` if they actually are cancelled

Reply:
> I'd rather not aggressively set task run states to Cancelled, only because if they do actually finish successfully it's more informative for their state to reflect that.

The logic I proposed above would still accomplish that though, since when a flow run transitions to `Cancelled`, there are no active task runs. I'd rather update this state at the same time, as I believe it provides a nicer user experience. Otherwise users will still see active tasks for a while in the UI before the zombie-killer eventually notices them.

Reply:
AH I understand now; yes I 100% agree - the backend should transition all non-Finished task runs to `Cancelled` whenever the flow run enters a `Cancelled` state instead of having the client do it 👍